### ISSUE TITLE
No Mastodon image descriptions for screen readers

### ISSUE THREAD

**Issue by alcapurrias** [Automattic/wordpress-activitypub#94]
Hello! I have created an Instagram-esque blog with the AP plugin, and have noticed that Mastodon is not reading anything for the image description. Image descriptions are strongly encouraged on there for accessibility reasons. 

I have filled out the alt text, title, description and caption but it doesn't carry over. Is this something that can be transmitted from Wordpress?

**Comment by wohfab** [comment]
I'd +1 this.

**Comment by mediaformat** [comment]
Confirming an issue with Image descriptions:

The ActivityPub representation of a post image includes the `alt` values from the Media Library, not those from the Image in the post, which can be different.

**Comment by timnolte** [comment]
I would sort of classify this as more of a WordPress Core issue. IMO WordPress shouldn't be having different `alt` between the Media Library and a post. I don't understand in what cases those should ever need to be different. As a matter of best practices the `alt` text should be set in the Media Library when uploading happens and then what I have seen is that it will be used on posts.

**Comment by pfefferle** [comment]
I think we have to change they way the plugin currently searches for images/attachments. Maybe try to use HTML API introduced in WP V6.2 https://make.wordpress.org/core/2023/03/07/introducing-the-html-api-in-wordpress-6-2/

**Comment by janboddez** [comment]
> As a matter of best practices the alt text should be set in the Media Library when uploading happens and then what I have seen is that it will be used on posts.

I agree, and I wouldn't even mind if an AP image's alt text differed a bit from what was in the post; as long as there's something.

The thing I recently ran into is that Gutenberg, or at least the Image block, does _not_ actually write alt text to the `wp_postmeta` table like the classic editor's upload modal does.

So you could grab _attached_ images (those uploaded through the Edit Post screens, and these do unfortunately _not_ necessarily equal the actual in-post images) and you could very well not find any alt text because, well, WordPress never stored it (outside the post content/block context).

What I did in my case is implement an alternative (to the kind of _attached_ images I just described), and that is a simple "scraper" that grabs the `src` and `alt` of (only) `img` elements, from the actual post content.

I did not use the HTML API (not sure if I should, as that would limit things to WP 6.2 and up). `DOMDocument`/`DOMXPath` work well enough (I think), although they can be tricky at first.

https://github.com/janboddez/share-on-mastodon/blob/ca3f18d63324aee8d17192b771dc23a829e95b7c/includes/class-image-handler.php#L90

**Comment by janboddez** [comment]
I want to give filtering `activitypub_post`, or, more specifically, the post array's `attachment` item, a try, to be able to also allow in-post images to be included, or at least fetch their alt tags, and try to map those back to the images in the list.

If that works, some of it could be merged back into the plugin, maybe.

**Comment by dogwelderx** [comment]
Mastodon strongly encourages alt text, and many users will ignore/delete/block comments without it. An ActivityPub plugin that doesn't include alt text is fundamentally broken.

**Comment by timnolte** [comment]
Just chiming in here that this is partly an aspect of broken functionality in WordPress Core. This is related to additionally broken functionality in the WordPress mobile app when uploading images to the Media Library and actually adding alt text to only have it never saved to the images in the Media Library at all. https://github.com/wordpress-mobile/WordPress-Android/issues/18514

**Comment by pfefferle** [comment]
what @timnolte says!

@dogwelderx if you are uploading images directly for the post and add alt-texts before including them into the post, they will be also used for ActivityPub: https://github.com/Automattic/wordpress-activitypub/blob/master/includes/model/class-post.php#L347

**Comment by dogwelderx** [comment]
Nope.
If I add a picture directly to the post, the picture appears in the fediverse post with no alt text. If add it from the media directory, I don't even get the picture.  YOu can see for yourself if you follow my site with your Fediverse account: ga2so_2wba65@www.ga2so.com

Wordpress is hosted on my own server, and I don't have Jetpack installed because my work's filter blocks images that are filtered through Jetpack. I'm guessing that's the issue.

**Comment by janboddez** [comment]
> If I add a picture directly to the post, the picture appears in the fediverse post with no alt text.

Sounds like Gutenberg/the block editor. Its alt text is not saved into the attachment's postmeta.

> If add it from the media directory, I don't even get the picture.

Attachments uploaded _not through the Edit/Create Post screen of the current post_ don't get the post as their `post_parent` and WordPress/the plugin doesn't fetch them even if they're added to the post later on. (WordPress does not consider them "attached images," which is a very confusing name.)

This behavior is what I described in https://github.com/Automattic/wordpress-activitypub/issues/94#issuecomment-1563927526. In the classic editor, it'd have worked.

Unless core addresses this, the only workaround I've been able to get to work most of the time, is parse the actual rendered post content for [images and their alt texts], and use that information to create the `attachment` array.
